scanner/e2etests/image_test.go

```diff
@@ -0,0 +1,244 @@
+}
+
+// TestWant describes the information wanted from the scan results in a TestCase.
+type TestWant struct {
+	Namespace string    `json:"namespace"`
+	Source    string    `json:"source"`
+	Features  []Feature `json:"expected_features"`
+}
+
+// TestArgs describes the arguments (i.e. test input) necessary to exercise the test case.
+type TestArgs struct {
+	Image    string `json:"image"`
+	Registry string `json:"registry"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// TestCase represents on test case or scenario.
+type TestCase struct {
+	TestArgs
+	TestWant
+	// Test flags.
+	OnlyCheckSpecifiedVulns  bool `json:"only_check_specified_vulns"`
+	UncertifiedRhel          bool `json:"uncertified_rhel"`
+	CheckProvidedExecutables bool `json:"check_provided_executables"`
+}
+
+// TestImage is the E2E test for container image scans.
+func TestImage(t *testing.T) {
+	ctx := context.Background()
+	c, err := client.NewGRPCScannerClient(ctx,
+		client.WithAddress(":8443"),
+		client.WithoutTLSVerify)
+	require.NoError(t, err)
+
+	testCases, err := loadTestCases("testdata/image_tests.json")
+	require.NoError(t, err)
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%s/%s", tc.Namespace, tc.Image), func(t *testing.T) {
+			ref, err := name.ParseReference(tc.Image)
+			require.NoError(t, err)
+
+			auth := tc.authConfig()
+			d, err := indexer.GetDigestFromReference(ref, auth)
+			require.NoError(t, err)
+
+			vr, err := c.IndexAndScanImage(ctx, d, auth)
+			require.NoError(t, err)
+
+			expected := &tc.TestWant
+			converted := tc.convertReport(vr)
+			defer func() {
+				if t.Failed() && vr != nil {
+					tc.logReport(t, vr)
+				}
+			}()
+			assert.Equal(t, expected, converted,
+				"The vulnerability report converted did not contain the expected values.")
+		})
+	}
+}
+
+func loadTestCases(s string) ([]*TestCase, error) {
+	f, err := os.Open(s)
+	if err != nil {
+		return nil, err
+	}
+	defer utils.IgnoreError(f.Close)
+	decoder := json.NewDecoder(f)
+	var cases []*TestCase
+	err = decoder.Decode(&cases)
+	if err != nil {
+		return nil, err
+	}
+	return cases, nil
+}
+
+// namespace creates a test case namespace from a vulnerability report.
+func namespace(vr *v4.VulnerabilityReport) string {
+	if len(vr.GetContents().GetDistributions()) == 1 {
+		d := vr.GetContents().GetDistributions()[0]
+		return fmt.Sprintf("%s:%s", d.GetDid(), d.GetVersionId())
+	}
+	return "unknown"
+}
+
+func (tc *TestCase) authConfig() authn.Authenticator {
+	return authn.FromConfig(authn.AuthConfig{
+		Username: os.Getenv(tc.Username),
+		Password: os.Getenv(tc.Password),
+	})
+}
+
+func (tc *TestCase) convertReport(vr *v4.VulnerabilityReport) *TestWant {
+	return &TestWant{
+		Namespace: namespace(vr),
+		Source:    tc.Source,
+		Features:  tc.convertFeatures(vr),
+	}
+}
+
+func (tc *TestCase) convertFeatures(vr *v4.VulnerabilityReport) []Feature {
+	nv := func(n, v string) string {
+		return fmt.Sprintf("%s-%s", n, v)
+	}
```

Comment:
another option to consider is to make the key a struct:

```
type featKey struct {
        name, version string
}

...


feats[featKey{f.Name, f.Version}]
```

accomplishes the same thing, though, so not a big deal. Just an option

Reply:
I don't fully remember, but having a string key yielded better logging/grepping.